scionlab/models/user_as.py

```diff
@@ -192,73 +127,183 @@ class Meta:
-               use_vpn,
-               public_ip,
-               public_port,
-               bind_ip,
-               bind_port,
                installation_type):
         """
         Update this UserAS instance and immediately `save`.
         Updates the related host, interface and link instances and will trigger
         a configuration bump for the hosts of the affected attachment point(s).
         """
-        prev_ap = self.attachment_point
-
-        host = self.hosts.get()   # UserAS always has only one host
-
-        if self.isd != attachment_point.AS.isd:
-            self._change_isd(attachment_point.AS.isd)
-
-        host.update(
-            public_ip=public_ip,
-            bind_ip=_VAGRANT_VM_LOCAL_IP if installation_type == UserAS.VM else bind_ip,
-        )
+        # XXX(andrea_tulimiero): Useless since the model instance has already got the updated fields
+        self.label = label
+        self.host.update()
 
-        link = self._get_ap_link()
-        interface_ap = link.interfaceA
-        interface_user = link.interfaceB
-        if use_vpn:
-            vpn_client = self._create_or_activate_vpn_client(attachment_point.vpn)
-            interface_user.update(
-                public_ip=vpn_client.ip,
-                public_port=public_port,
-                bind_port=None
-            )
-            interface_ap.update(
-                border_router=attachment_point.get_border_router_for_useras_interface(),
-                public_ip=attachment_point.vpn.server_vpn_ip,
-                public_port=None,
-            )
-        else:
-            host.vpn_clients.update(active=False)   # deactivate all vpn clients
-            interface_user.update(
-                public_ip=None,
-                public_port=public_port,
-                bind_ip=None,
-                bind_port=bind_port
-            )
-            interface_ap.update(
-                border_router=attachment_point.get_border_router_for_useras_interface(),
-                public_ip=None,
-                public_port=None,
-            )
-        self.attachment_point = attachment_point
-        self.installation_type = installation_type
-        self.public_ip = public_ip
-        self.bind_ip = bind_ip
-        self.bind_port = bind_port
         self.save()
 
-        if self.attachment_point != prev_ap:
-            prev_ap.split_border_routers()  # clean up empty BRs
-            prev_ap.trigger_deployment()
-        self.attachment_point.split_border_routers()
-        self.attachment_point.trigger_deployment()
+    @property
+    def host(self):
+        return self.hosts.get()  # UserAS always has only one host
 
+    def is_link_over_vpn(self, link):
+        """
+        Returns whether the link is over vpn
+        """
+        ap = link.interfaceA.AS.attachment_point_info
+        if ap.vpn is None:
+            return False
+        vpn_client = VPNClient.objects.filter(host=self.host, vpn=ap.vpn).first()
```

Comment:
Yeah, way cleaner. Although `isinstance(...)` should fail, since traversing `interface.AS` will always return an instance of `AS`. Nevertheless, we can check this with `hasattr(interface.AS, 'useras')`